src/v/cluster/jumbo_log_stm.cc

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "base/vassert.h"
+#include "cluster/jumbo_log_stm.h"
+
+#include "base/vlog.h"
+#include "bytes/iobuf.h"
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "errc.h"
+#include "jl/metadata.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "model/record_batch_reader.h"
+#include "model/record_batch_types.h"
+#include "model/timeout_clock.h"
+#include "persisted_stm.h"
+#include "raft/consensus.h"
+#include "replicate.h"
+#include "serde/envelope.h"
+#include "serde/rw/rw.h"
+#include "storage/record_batch_builder.h"
+#include "utils/named_type.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/util/bool_class.hh>
+#include <seastar/util/later.hh>
+#include <seastar/util/log.hh>
+
+#include <__utility/move.h>
+#include <fmt/core.h>
+
+#include <stdexcept>
+#include <stdint.h>
+#include <system_error>
+#include <tuple>
+
+namespace cluster {
+
+struct jumbo_log_stm::write_intent
+  : public serde::
+      envelope<write_intent, serde::version<0>, serde::compat_version<0>> {
+    ::jumbo_log::write_intent_segment segment;
+};
+
+using cmd_key = named_type<uint8_t, struct cmd_key_tag>;
+struct jumbo_log_stm::create_write_intent_cmd {
+    static constexpr cmd_key key{0};
+    using value = jumbo_log_stm::write_intent;
+};
+
+jumbo_log_stm::jumbo_log_stm(ss::logger& logger, raft::consensus* c)
+  : jumbo_log_stm(logger, c, config::shard_local_cfg()) {}
+
+jumbo_log_stm::jumbo_log_stm(
+  ss::logger& logger, raft::consensus* c, config::configuration& cfg)
+  : persisted_stm(jumbo_log_snapshot, logger, c)
+  , _next_write_intent_id(1) {
+    std::ignore = cfg;
+}
+
+// TODO(nv): Fix all raft::make_error_code calls to use proper error codes.
+//           Also, consider using std::error_code instead of std::system_error.
+
+ss::future<jumbo_log::write_intent_id_t> jumbo_log_stm::create_write_intent(
+  jumbo_log::segment_object object, model::timeout_clock::duration timeout) {
+    // TODO(nv): this lock needs a timeout
+    auto u = co_await _lock.get_units();
+
+    if (!co_await sync(timeout)) {
+        throw std::system_error(raft::make_error_code(raft::errc::timeout));
```

Comment:
Most likely. I believe I added a todo above to review this. Used exceptions as a shortcut to avoid too verbose code in the caller in the interest of time.

Reply:
done